### PR TITLE
[CDAP-20989] Add API implementations for Namespace and Repository Source Control Metadata

### PR DESCRIPTION
**Pull request by adrikagupta**
- Implemented APIs to get namespace source control metadata 
- Implemented API to retrieve repository source control metadata
- Added Pagination, filtering (based on name, sync status), sorting (based on name and last modified)
- Added unit tests for the above

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ApplicationFilter.java [hunk 1]
```diff
@@ -47,7 +47,7 @@ public abstract static class ArtifactIdFilter extends ApplicationFilter implemen
       Predicate<ArtifactId> {
 
   }
-
+
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ApplicationFilter.java]: Please revert

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanApplicationsRequest.java [hunk 2]
```diff
@@ -70,7 +70,6 @@ private ScanApplicationsRequest(@Nullable NamespaceId namespaceId,
     this.latestOnly = latestOnly;
     this.sortCreationTime = sortCreationTime;
   }
-
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanApplicationsRequest.java]: Please revert

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+package io.cdap.cdap.app.store;
+
+import io.cdap.cdap.proto.sourcecontrol.SortBy;
+import io.cdap.cdap.proto.sourcecontrol.SortBy;
+import io.cdap.cdap.spi.data.SortOrder;
+
+public class ScanSourceControlMetadataRequest {
+  private final String namespace;
+  private final SortOrder sortOrder;
+  private final SortBy sortOrderOn;
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java]: sortOn

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.app.store;
+
+import io.cdap.cdap.proto.sourcecontrol.SortBy;
+import io.cdap.cdap.proto.sourcecontrol.SortBy;
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java]: duplicate import?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java [hunk 5]
```diff
@@ -82,6 +84,10 @@ public interface Store {
   void setProvisioning(ProgramRunId id, Map<String, String> runtimeArgs,
                        Map<String, String> systemArgs, byte[] sourceId, ArtifactId artifactId);
 
+  void cleanupRepoSourceControlMeta(String namespace, HashSet<String> repoFiles);
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java]: Dont use default store for these. We can directly use `SourceControlMetadataStore`s

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 6]
```diff
@@ -287,6 +294,78 @@ public void getAllApps(HttpRequest request, HttpResponder responder,
...
+    JsonPaginatedListResponder.respond(GSON, responder, APP_LIST_PAGINATED_KEY,
+        jsonListResponder -> {
+          AtomicReference<SourceControlMetadataRecord> lastRecord = new AtomicReference<>(null);
+          ScanSourceControlMetadataRequest scanRequest = SourceControlMetadataHelper.getScmStatusScanRequest(
+              namespaceId,
+              pageToken, pageSize, sortOrder, sortOrderOn, filter);
+          boolean pageLimitReached = false;
+          try {
+            pageLimitReached = applicationLifecycleService.scanSourceControlMetadata(
+                scanRequest, 20,
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: why 20 ?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 7]
```diff
@@ -287,6 +294,78 @@ public void getAllApps(HttpRequest request, HttpResponder responder,
...
+                scmMetaRecord -> {
+                  SourceControlMetadataDetail detail = new SourceControlMetadataDetail(
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: I think we should just send `SourceControlMeta`. UI can decide what to show

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java [hunk 8]
```diff
@@ -392,6 +397,56 @@ public RepositoryAppsResponse listApps(NamespaceId namespace) throws NotFoundExc
...
   }
 
+  public boolean scanRepoMetadata(ScanSourceControlMetadataRequest request, int txBatchSize,
+      Consumer<SourceControlMetadataRecord> consumer) throws NotFoundException,
+      AuthenticationConfigException, IOException {
+    NamespaceId namespaceId = new NamespaceId(request.getNamespace());
+    accessEnforcer.enforce(namespaceId, authenticationContext.getPrincipal(),
+        NamespacePermission.READ_REPOSITORY);
+    RepositoryConfig repoConfig = getRepositoryMeta(namespaceId).getConfig();
+    RepositoryAppsResponse repositoryAppsResponse = sourceControlOperationRunner.list(
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java:407]: This should be done in a background job

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 7 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: SourceControlMeta doesn't have the file name. It has file hash, commitId, LastSynced, and syncStatus. Should I add another property 'name'?

Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Oh right. In that case this is ok. Please add javadoc in the class to explain what is the usecase

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ApplicationFilter.java (hunk 1 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ApplicationFilter.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanApplicationsRequest.java (hunk 2 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanApplicationsRequest.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java (hunk 4 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java (hunk 5 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 6 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 7 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Sure. Added in SourceControlMetadataDetail class

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java (hunk 8 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java:407]: Added a comment for it

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java (hunk 3 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/ScanSourceControlMetadataRequest.java]: done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+ * the License.
+ */
+
+package io.cdap.cdap.proto;
+
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Represents a record of source control metadata.
```
Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java:23]: Please add more details. Explanin how this is different from `SourceControlMeta`

File: cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  private final String namespace;
+  private final String type;
+  private final String name;
+  private final String specificationHash;
+  @Nullable
+  private final String commitId;
+  @Nullable
+  private final Long lastModified;
+  @Nullable
+  private final Boolean isSynced;
```
Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java:37]: This should not be nullable

File: cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataDetail.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.proto;
+
+/**
+ * This class is used for responses of API calls that fetch source control metadata.
+ */
+public class SourceControlMetadataDetail {
```
Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataDetail.java]: Why do we have `SourceControlMetadataDetail` and `SourceControlMetadataRecord`. We could just return `SourceControlMetadataRecord`

File: cdap-common/src/main/java/io/cdap/cdap/common/SourceControlMetadataNotFoundException.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+
+/**
+ * Exception thrown when source control metadata for an application is not found.
+ */
+public class SourceControlMetadataNotFoundException extends NotFoundException {
+
+  private final ApplicationReference appRef;
+
+  public SourceControlMetadataNotFoundException(ApplicationReference appRef) {
```
Comment (samdgupi) [inline cdap-common/src/main/java/io/cdap/cdap/common/SourceControlMetadataNotFoundException.java:34]: Add javadoc

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStoreTest.java [hunk 13]
```diff
@@ -33,8 +45,239 @@ public abstract class RepositorySourceControlMetadataStoreTest {
   private static final ApplicationReference APP_REF = new ApplicationReference(NAMESPACE, NAME);
   private static final Boolean IS_SYNCED = true;
 
+  private static final String TYPE = EntityType.APPLICATION.toString();
+
   protected static TransactionRunner transactionRunner;
 
+  @Test
+  public void testScanRepositorySourceControlMeta() {
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStoreTest.java]: divide into seperatev tests for normal/sort/filter

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlRepositorySourceControlMetadataStoreTest.java [hunk 14]
```diff
@@ -33,6 +34,11 @@ public static void beforeClass() throws Exception {
     transactionRunner = injector.getInstance(TransactionRunner.class);
   }
 
+  @Override
+  @Test
+  public void testScanRepositorySourceControlMeta() {
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlRepositorySourceControlMetadataStoreTest.java]: Please add a comment on why it does not work and why it is empty

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlNamespaceSourceControlMetadataTest.java [hunk 15]
```diff
@@ -32,6 +34,12 @@ public static void beforeClass() throws Exception {
     transactionRunner = injector.getInstance(TransactionRunner.class);
   }
 
+  // TODO (CDAP-21005): Currently, implementation for sorting on strings is broken for NoSQL StructuredTable
+  @Override
+  @Test
+  public void testScanNamespaceSourceControlMeta() {
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlNamespaceSourceControlMetadataTest.java]: same comment as RepositorySourceControlMeta

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java [hunk 16]
```diff
@@ -33,29 +45,269 @@ public abstract class NamespaceSourceControlMetadataStoreTest {
...
-      COMMIT_ID, LAST_MODIFIED, true);
+      COMMIT_ID, LAST_MODIFIED, IS_SYNCED);
 
   protected static TransactionRunner transactionRunner;
 
   @Before
   public void before() {
     TransactionRunners.run(transactionRunner, context -> {
-      NamespaceSourceControlMetadataStore store = NamespaceSourceControlMetadataStore.create(context);
+      NamespaceSourceControlMetadataStore store = NamespaceSourceControlMetadataStore.create(
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java]: Please revert back all reformatting that is not needed

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java [hunk 17]
```diff
@@ -33,29 +45,269 @@ public abstract class NamespaceSourceControlMetadataStoreTest {
...
+          context);
       store.write(APP_REF, SOURCE_CONTROL_META);
     });
   }
 
+  @Test
+  public void testScanNamespaceSourceControlMeta() {
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java]: Divide into seperate tests for sort/filter

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java [hunk 18]
```diff
@@ -142,12 +143,12 @@ public ApplicationLifecycleService createLifeCycleService(CConfiguration cConf,
           MetadataServiceClient metadataServiceClient,
           AccessEnforcer accessEnforcer, AuthenticationContext authenticationContext,
           MessagingService messagingService, Impersonator impersonator,
-          CapabilityReader capabilityReader) {
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java:145]: what is this change used for ?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java [hunk 19]
```diff
@@ -58,14 +70,87 @@ private StructuredTable getRepositorySourceControlMetadataTable() {
...
+            request.getNamespace())
+    );
+    startFields.add(
+        Fields.stringField(StoreDefinition.RepositorySourceControlMetadataStore.TYPE_FIELD,
+            type)
+    );
+
+    if (request.getFilter().getSyncStatus() != null) {
+      startFields.add(
+          Fields.booleanField(StoreDefinition.RepositorySourceControlMetadataStore.IS_SYNCED_FIELD,
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java:102]: If we are using enums as API filters we might want to use the same for DB. Is there a specific reason why we decided upon filter to be an enum rather than boolean ?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java [hunk 20]
```diff
@@ -58,14 +70,87 @@ private StructuredTable getRepositorySourceControlMetadataTable() {
...
+
   /**
    * Writes or updates the source control metadata for the specified repository metadata ID.
    *
-   * @param appRef The {@link ApplicationReference} for which source control metadata needs to be
-   *                       written or updated.
-   * @param isSynced       A boolean indicating whether the entity in repository is synced with the
-   *                       one in namespace
-   * @param lastSynced     The timestamp of the last git pull/push of the entity
+   * @param appRef     The {@link ApplicationReference} for which source control metadata needs to
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java:147]: revert this change please

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java [hunk 21]
```diff
@@ -58,14 +70,87 @@ private StructuredTable getRepositorySourceControlMetadataTable() {
...
+          Fields.booleanField(StoreDefinition.RepositorySourceControlMetadataStore.IS_SYNCED_FIELD,
+              request.getFilter().getSyncStatus().toString().equals("SYNCED")));
+    }
+
+    Collection<Field<?>> endFields = startFields;
+
+    if (request.getScanAfter() != null) {
+      startBound = Range.Bound.EXCLUSIVE;
+      startFields =
+          request.getSortOn() == SortBy.LAST_SYNCED_DATE ? getLastModifiedRangeFields(request, type)
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java]: why not have one function that takes sorton as parameter

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java [hunk 22]
```diff
@@ -969,6 +985,57 @@ public ApplicationMeta getLatest(ApplicationReference appRef) {
...
+      Consumer<SourceControlMetadataRecord> consumer) {
+    return TransactionRunners.run(transactionRunner, context -> {
+      String type = EntityType.APPLICATION.toString();
+      return getRepoSourceControlMetadataStore(context).scan(request, type,
+          consumer);
+    });
+  }
+
+  @Override
+  public int scanNamespaceSourceControlMetadata(ScanSourceControlMetadataRequest request,
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java]: name it scanAppSourceControlMetadata as Default store should contain all app related functionalities

File: cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataDetail.java (hunk 11 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataDetail.java]: Yes, we could but `SourceControlMetadataRecord` has a lot of extra fields. Hence, I had created `SourceControlMetadataDetail`. I'll make the changes since it is fine to return the record

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java (hunk 18 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java:145]: We are using transaction runner in ApplicationLifecycleService for scan method. Here we are creating an instance of ApplicationLifecycleservice for injection and it needs transaction runner as one of the args

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java (hunk 20 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java:147]: Actually, this is a fix of checkstyle indentation warning

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java (hunk 22 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java]: Sure, done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java (hunk 21 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java]: Agreed. Done. Thanks for the correction

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java (hunk 19 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStore.java:102]: There's no specific reason for that, we could have boolean as well. I have made the change

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java (hunk 17 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java]: sure, done

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java (hunk 16 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStoreTest.java]: sure, done

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlNamespaceSourceControlMetadataTest.java (hunk 15 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlNamespaceSourceControlMetadataTest.java]: noted

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlRepositorySourceControlMetadataStoreTest.java (hunk 14 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/NoSqlRepositorySourceControlMetadataStoreTest.java]: sure, done

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStoreTest.java (hunk 13 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/RepositorySourceControlMetadataStoreTest.java]: sure, done

File: cdap-common/src/main/java/io/cdap/cdap/common/SourceControlMetadataNotFoundException.java (hunk 12 above)
Comment (adrikagupta) [reply cdap-common/src/main/java/io/cdap/cdap/common/SourceControlMetadataNotFoundException.java:34]: sure, done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataDetail.java (hunk 11 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataDetail.java]: done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java (hunk 10 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java:37]: made the necessary changes

File: cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java (hunk 9 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/SourceControlMetadataRecord.java:23]: sure, done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/util/SourceControlMetadataHelper.java [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+      builder.setSortOn(sortOn);
+    }
+    if (filter != null && !filter.isEmpty()) {
+      builder.setFilter(getFilter(filter));
+    }
+
+    return builder.build();
+  }
+
+  private static SourceControlMetadataFilter getFilter(String filterStr)
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/util/SourceControlMetadataHelper.java:74]: Please add unit test for this

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java [hunk 24]
```diff
@@ -767,7 +783,7 @@ public void removeApplication(ApplicationReference appRef) {
   @Override
   public void removeApplication(ApplicationId id) {
     LOG.trace("Removing application: namespace: {}, application: {}", id.getNamespace(),
-              id.getApplication(), id.getVersion());
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java:770]: please revert

**Review by samdgupi — APPROVED** [review]
